[cmake] update + build crossguid for all platforms internally by default

fuzzard opened the pull request:
## Description
Bump crossguid to current master tip hash [ca1bf4b810e2d188d04cb6286f957008ee1b7681](https://github.com/graeme-hill/crossguid/tree/ca1bf4b810e2d188d04cb6286f957008ee1b7681)

## Motivation and context
Our previous crossguid hash was from 2015. theres been a number of updates to crossguid, but no actual release since August 2017, so we'll just pick current latest hash

Big changes are, crossguid now natively builds using cmake. We can then remove our CMakeLists.txt used for internal crossguid.
Crossguid now requires C++17, which we now fullfill !
crosssguid defaults to setting the flag -Werror. Unfortunately this causes build failures on osx (at least) https://github.com/graeme-hill/crossguid/issues/48
The patch gets around this build failure using a c++17 attribute

I did test the codepath for ENABLE_INTERNAL_CROSSGUID, so that should be fine as part of this.

## How has this been tested?
build osx

## What is the effect on users?
N/A

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[cmake] update + build crossguid all platforms

crossguid hasnt been updated in a long time (~2015 was our previous hash)
crossguid now natively uses cmake, so remove our cmakelists file

first patch is to get around a failure due to unused function on apple platforms.
a second patch is introduced due to a build failure with the default -Werror used

Build crossguid in core cmake project for all platforms by default.
Remove from windows the bundled download of crossguid lib

rework usage of debug_postfix in SETUP_BUILD_VARS macro

fuzzard commented:
OK, so ive reworked this to build crossguid for every platform by default. removed existing libs from windows, tools/depends no longers builds crossguid.

Got hung up on some packaging changes required for windows, so figure maybe this is a suitable way to not have to deal with that @Paxxi 
Let me know if you want any patches brought in for win platforms. or even if you dont want to go this route.

fuzzard commented:
So, due to jenkins build server access issues, ive implemented a FindPatch module to handle downloading a patch binary from our mirrors. The bulk of that is based off cmake/scripts/windows/tools/patch.cmake that is used for addon building.

Ideally we'd get the specific jenkins builder (Win-Montellese) to have an appropriate patch binary installed, but this sidesteps the access issues currently with that.

Ill ping @Paxxi to see if this is acceptable from the windows side of things.

Paxxi reviewed (approved):
Looks good to me

Paxxi commented:
One thing I thought of, with visual studio we don't set any build type during project generation. We can switch between release and debug builds, does this handle that so both crossguid and crossguidd are built and available?

fuzzard commented:
Currently I've taken a shortcut for crossguid in this PR. We just use release for all build types.
I have another pr #21043 that handles such build/switching. My intent is to make all internal deps to be able to built and work with libs using debug postfixes for multiconfig generators (VS and Xcode). Just baby steps for now. I find it easier to get things merged/reviewed if I go slowly/incrementally

Commit:
[cmake] FindFMT relocate select_library_configurations call

relocate select_library_configurations to generic code path. Multi Config generators (Xcode, VS)
will be able to autodetect debug/release lib names for internal builds

Commit:
[cmake] FindSpdlog relocate select_library_configurations call

relocate select_library_configurations to generic code path. Multi Config generators (Xcode, VS)
will be able to autodetect debug/release lib names for internal builds

Commit:
[cmake] rework DEBUG_POSTFIX usage in SETUP_BUILD_VARS macro

fuzzard commented:
ok @Paxxi revisited this for that use case.
Of note, with crossguid they use an "unusual" debug_postfix (-dgb). Instead of patching this out to a more standard d postfix, ive allowed a dep build to set the postfix, or use the "global" d postfix. Atm ive only continued to set the global postfix for multiconfig generators (VS/Xcode). As mentioned about my other PR regarding non hardcoded build types, i'll handle the unix makefiles case in that PR.
For a lib that doesnt use a postfix for debug, it will also allow us to override if so desired.
```
set(<MODULE>_DEBUG_POSTFIX "")
```

So, for multiconfig generators (VS and xcode) when you choose your build type

eg for VS
```
cmake --build . --config "Debug"
```

or xcode
```
xcodebuild -configuration Debug build
```
and each scheme/target will build the lib (or use if already built) relevant to the type. On first run, it only builds for whatever build type is requested.

fuzzard commented:
It's not Forced per se, the option is defaulted on, so if you override with NO, it'll still look for a system provided crossguid and if there is none, will fail

howie-f commented:
then something is not working like intended because
`ENABLE_INTERNAL_CROSSGUID=NO` and
```
ii  libcrossguid-dev:amd64        0.0+git200150803-4             amd64        C++ UUID library headers
ii  libcrossguid0:amd64           0.0+git200150803-4             amd64        C++ UUID library
```
fails after this pr
```
CMake Error at /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:545 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:733 (_pkg_check_modules_internal)
  cmake/modules/FindCrossGUID.cmake:53 (pkg_check_modules)
  cmake/scripts/common/Macros.cmake:372 (find_package)
  cmake/scripts/common/Macros.cmake:386 (find_package_with_ver)
  CMakeLists.txt:190 (core_require_dep)
```

howie-f commented:
if i revert the 4 commits i can detect the distro shipped version again
`-- Found CrossGUID: /usr/lib/x86_64-linux-gnu/libcrossguid.so`

fuzzard commented:
Shoot me your environment details on Slack and I'll run up a repro environment and look into it

lrusak commented on cmake/modules/FindCrossGUID.cmake:
```diff
@@ -1,43 +1,85 @@
 
+  # Temp: We force CMAKE_BUILD_TYPE to release, and makefile builds respect this
+  # Multi config generators (eg VS, Xcode) dont, so handle debug postfix build/link for them only
+  if(NOT CMAKE_GENERATOR STREQUAL "Unix Makefiles")
```
This probably isn't quite right as the ninja generator should be considered similar to Unix makefiles.

fuzzard replied:
added a check for ninja generator in #21178

lrusak commented:
the pkg-config was only added in the very last commit of crossguid https://github.com/graeme-hill/crossguid/commit/ca1bf4b810e2d188d04cb6286f957008ee1b7681

Also,
```
~$ rpm -ql crossguid                 
/usr/lib/.build-id
/usr/lib/.build-id/12
/usr/lib/.build-id/12/5208117e28e864f4fef1d3a6451ac21c22230d
/usr/lib64/libcrossguid.so.0
/usr/lib64/libcrossguid.so.0.0
/usr/share/doc/crossguid
/usr/share/doc/crossguid/README.md
/usr/share/licenses/crossguid
/usr/share/licenses/crossguid/LICENSE

~$ rpm -ql crossguid-devel           
/usr/include/guid.h
/usr/lib64/libcrossguid.so
```

fuzzard commented:
what version does it say crossguid is for you? 0.2.0?

The issue we have with crossguid (and rapidjson is in the exact same boat) is upstream just wont tag anything newer. Theres been dev on both for years, without a tagged release since like before 2017.

the cmake introduction of crossguid isnt in any tagged release, but has been in tree since like 2017. They reworked location of headers to be in package named folders as well.

Honestly, i dont know how to handle that (particularly the includes in our code for the headers). I guess more defines based on path to the header in the find_include. i dunno. This is all just far harder because upstream will not do a tag, therefore distros wont package/ship anything newer than ancient.

lrusak commented:
> what version does it say crossguid is for you? 0.2.0?

```
$ rpm -qa crossguid                                                                                                                                                                                                         
crossguid-0-0.17.20160908gitfef89a4.fc35.x86_64
```

also, https://src.fedoraproject.org/rpms/crossguid

fuzzard commented:
In that case, it should resort to not defining HAVE_NEW_CROSSGUID, and use the ifdef at https://github.com/xbmc/xbmc/blob/a508a18d16f801c89dbb6201f0869da9db580470/xbmc/utils/StringUtils.cpp#L19-L23 

I'll revert the pkg-config usage, or maybe look at a fallback if pkg-config vars arent found. I'll see what works best.
